### PR TITLE
Add type checking in CI and fixup annotations

### PR DESCRIPTION
**Pull request by PeterJCLaw**
This adds type checking in CI with `mypy`, a basic configuration for `mypy` and fixes the errors that are found.

The configuration is very nearly the default and mostly doesn't enable additional checks. The checks which are enabled are mostly to guide future usage in the right direction.

The fixes here are intended to be as minimal as possible and are derived from code inspection. I've run the tests, but don't have an actual deployment to test this with manually.

### CODE REVIEW

File: pythonosc/osc_server.py [hunk 1]
```diff
@@ -44,18 +49,18 @@ def _is_valid_request(request: List[bytes]) -> bool:
...
+    def __init__(self, server_address: Tuple[str, int], dispatcher: Dispatcher, bind_and_activate: bool = True) -> None:
         """Initialize
 
         Args:
             server_address: IP and port of server
             dispatcher: Dispatcher this server will use
             (optional) bind_and_activate: default=True defines if the server has to start on call of constructor  
         """
-        super().__init__(server_address, _UDPHandler, bind_and_activate)
+        super().__init__(server_address, _UDPHandler, bind_and_activate)  # type: ignore[call-arg]  # https://github.com/python/typeshed/pull/8542
```
Comment (attwad) [inline pythonosc/osc_server.py:60]: this PR has been merged now, please fix

**Review by attwad — CHANGES_REQUESTED** [review]

File: pythonosc/osc_server.py (hunk 1 above)
Comment (PeterJCLaw) [reply pythonosc/osc_server.py:60]: It has, however the typeshed is distributed within `mypy` so this is blocked until the next `mypy` release.
Once a version of `mypy` with the fix is used we should get a warning that the ignore is no longer valid which will be trivial to fix. Currently this code still passes using the latest `mypy`.

Comment (attwad) [reply pythonosc/osc_server.py:60]: okay thanks!

**Comment by attwad** [comment]
Checks on 3.6 seem to fail, can you please take a look?

**Comment by PeterJCLaw** [comment]
> Checks on 3.6 seem to fail, can you please take a look?

Huh, that actually looks like a legitimate issue -- on Python 3.6, [`asyncio.run`](https://docs.python.org/3/library/asyncio-task.html#asyncio.run) doesn't exist yet (it's introduced in Python 3.7). How would you like to handle that?

One option could be to have the example code do something like:
``` python
if sys.version_info < (3, 7):
    asyncio.run(...)
else:
    # TODO(python-upgrade): drop this once 3.6 is no longer supported
    loop = asyncio.get_event_loop()
    loop.run_until_complete(...)
    loop.close()
```

**Comment by attwad** [comment]
> > Checks on 3.6 seem to fail, can you please take a look?
> 
> Huh, that actually looks like a legitimate issue -- on Python 3.6, [`asyncio.run`](https://docs.python.org/3/library/asyncio-task.html#asyncio.run) doesn't exist yet (it's introduced in Python 3.7). How would you like to handle that?
> 
> One option could be to have the example code do something like:
> 
> ```python
> if sys.version_info < (3, 7):
>     asyncio.run(...)
> else:
>     # TODO(python-upgrade): drop this once 3.6 is no longer supported
>     loop = asyncio.get_event_loop()
>     loop.run_until_complete(...)
>     loop.close()
> ```

Thanks for looking into it, we can also just get rid of that pygame example, I haven't run it in ages and it adds yet another dependency that I'd prefer not having.

**Comment by PeterJCLaw** [comment]
> Thanks for looking into it, we can also just get rid of that pygame example, I haven't run it in ages and it adds yet another dependency that I'd prefer not having.

I'm not sure that would help here? The pygame example is fine (now; I realise I'd broken it when you first looked), it's `examples/async_server.py` which has the issue with `asyncio.run`.

**Comment by attwad** [comment]
okay let's do the condition check you mentioned, I'll see if dropping 3.6 support makes sense later.

**Review by attwad — APPROVED** [review]

**Pull request merged**